### PR TITLE
Add check to avoid overriding existing keys

### PR DESCRIPTION
**Pull request by BlaineHeffron**
### What

`stellar keys generate` will no longer overwrite an identity if it already exists, unless a different seed is provided.

### Why

Workflows that  utilize `generate` might unintentionally overwrite during contract development. 

Close https://github.com/stellar/stellar-cli/issues/1380

### Known limitations

N/A

### CODE REVIEW

File: cmd/soroban-cli/src/commands/keys/generate.rs
Comment (willemneal) [inline cmd/soroban-cli/src/commands/keys/generate.rs:132]: Too hard to write inline suggestion, but I would use the following since you don't care what the error is.
```suggestion
        let Some(existing_secret) = self.config_locator.read_identity(&self.name) {
        ...
        } else {
             // Identity doesn't exist, create new one
            self.write_and_fund_identity(&secret).await?;
        };
    
```

File: cmd/soroban-cli/src/commands/keys/generate.rs
Comment (willemneal) [inline cmd/soroban-cli/src/commands/keys/generate.rs:105]:
```suggestion
        let Ok(existing_secret) = self.config_locator.read_identity(&self.name) else {
            // Identity doesn't exist, create new one
            self.write_and_fund_identity(&secret).await?;
            return Ok(());
        };
```
Then  the stuff currently indented in the `if` can be shifted over for better readability.

**Review by leighmcculloch — CHANGES_REQUESTED** [review]
We can't merge this change at the moment, not until closer to a v22 release, because it breaks existing behavior.

**Comment by janewang** [comment]
Hi @BlaineHeffron Did we agree on this feature? Could you show me the thread of the discussion please

**Comment by willemneal** [comment]
@leighmcculloch How about a flag to prevent the overwrite? Then we can deprecate it in the next breaking release?

**Comment by BlaineHeffron** [comment]
> Hi @BlaineHeffron Did we agree on this feature? Could you show me the thread of the discussion please

There isn't a discussion thread that I'm aware of, it was something we had discussed with other AhaLabs team members.

**Comment by leighmcculloch** [comment]
> @leighmcculloch How about a flag to prevent the overwrite? Then we can deprecate it in the next breaking release?

It's unclear how effective an option would be. I doubt the average CLI user is going to start using an `--prevent-overwrite` option when generating keys, it just seems a bit esoteric.

What use case / workflow does the option serve?

**Comment by BlaineHeffron** [comment]
The app `Loam` which we are developing has a configuration option to auto generate contract wasms and install / deploy them in addition to generating and funding accounts. Currently we have a workaround where we check if the account exists before running `generate`. The problem with overwriting the account is someone might have some contract state tied to a specific account which would get lost if the account is overwritten.

**Comment by chadoh** [comment]
Original discussion thread in #1380 and [in Slack](https://aha-labs.slack.com/archives/C04B02ABF37/p1718654369768779). @janewang, @fnando, @elizabethengelman, and @willemneal were all in strong agreement with "overwrite by default is a bug" back in June.

**Comment by leighmcculloch** [comment]
> ...were all in strong agreement with "overwrite by default is a bug" back in June.

Yup, I think the idea of avoiding overwriting a key, which is a destructive action, is good too, but it needs to be held until the next major version because it is a breaking change. And I was otherwise only pushing back on the suggestion to add a temporary `--no-overwrite` since a major release is planned in the near future.

@Ifropc has labelled the issue as a `breaking change` meaning we'll circle back to it to the period leading up to the major release when main is open for breaking changes.

File: cmd/soroban-cli/src/commands/keys/generate.rs
Comment (leighmcculloch) [inline cmd/soroban-cli/src/commands/keys/generate.rs:124]: I think the idea that the command will implicitly overwrite in some edge cases is surprising. If we're saying "overwrite by default is a bug" then it's a bug that it still overwrites here.

The only time overwriting should be allowed is if an explicit `--overwrite` is provided so that the overwrite only occurs explicitly.

File: cmd/soroban-cli/src/commands/keys/generate.rs
Comment (leighmcculloch) [inline cmd/soroban-cli/src/commands/keys/generate.rs:88]: If someone attempts to generate a key for an alias that already has a key, if the command won't be doing it's primary function of generating a new key, we should have the command error with a non-zero exit status. Otherwise the command appears to have created a new-random-never-seen-before key, when in fact the alias may refer to a key that's not new and not random, which could be a security concern.

File: cmd/soroban-cli/src/commands/keys/generate.rs
Comment (leighmcculloch) [inline cmd/soroban-cli/src/commands/keys/generate.rs:90]: I don't think the command should fund an existing account. Unclear the purpose this serves since the generate command is primarily about getting a new key. If someone needs to top up their account they can use the fund command.

There's an open issue to remove funding from the generate command:
- https://github.com/stellar/stellar-cli/issues/1407

**Review by leighmcculloch — COMMENTED** [review]
Couple more things I noticed that I think we should change before merging this in the next major release.

File: cmd/soroban-cli/src/commands/keys/generate.rs
Comment (BlaineHeffron) [reply cmd/soroban-cli/src/commands/keys/generate.rs]: I added an `--overwrite` flag which simplified the logic as there was no need to compare existing secrets.

File: cmd/soroban-cli/src/commands/keys/generate.rs
Comment (BlaineHeffron) [reply cmd/soroban-cli/src/commands/keys/generate.rs]: I now throw an error `Error::IdentityAlreadyExists` which gets thrown if they try to run it without overwrite but the identity is found.

File: cmd/soroban-cli/src/commands/keys/generate.rs
Comment (BlaineHeffron) [reply cmd/soroban-cli/src/commands/keys/generate.rs]: I removed --no-fund and the funding portions.

**Comment by BlaineHeffron** [comment]
Comments have been addressed. This PR now also resolves https://github.com/stellar/stellar-cli/issues/1407

- I added an `--overwrite` flag to force overwrites. Otherwise it never overwrites. 
- If it finds that it would overwrite but `--overwrite` is not set, it throws an Error.
- It no longer funds the account afterwards. This also means there is no longer a `--no-fund` arg. 

Now I need to fix all the integration tests so that they fund the account after generate

**Comment by BlaineHeffron** [comment]
Actually, does it make sense to just add a --fund so that you could do both at once like the old functionality? (Would also make it easier for me to fix the tests ) @leighmcculloch

**Comment by BlaineHeffron** [comment]
Tests are fixed, should be ready.

**Comment by fnando** [comment]
Closing in favor of #1709

**Pull request closed without merging**
